server/src/test/java/org/elasticsearch/action/ingest/WriteableIngestDocumentTests.java

```diff
@@ -147,4 +151,71 @@ public void testToXContent() throws IOException {
+        WriteableIngestDocument testInstance = new WriteableIngestDocument(RandomDocumentPicks.randomIngestDocument(random(), true));
+        XContentType xContentType = randomFrom(XContentType.values());
+        BytesReference shuffled = toShuffledXContent(testInstance, xContentType, ToXContent.EMPTY_PARAMS, false);
+        XContentParser parser = createParser(XContentFactory.xContent(xContentType), shuffled);
+        WriteableIngestDocument parsed = doParseInstance(parser);
+        assertTrue(
+            testSourceMapEquality(
+                testInstance.getIngestDocument().getSourceAndMetadata(),
+                parsed.getIngestDocument().getSourceAndMetadata()
+            )
+        );
+    }
+
+    private boolean testSourceMapEquality(Map<String, Object> first, Map<String, Object> second) {
+        if (first == null) {
+            return second == null;
+        } else if (second == null) {
+            return false;
+        } else if (first.size() != second.size()) {
+            return false;
+        } else {
+            for (Map.Entry<String, Object> entry: first.entrySet()) {
+                Object value = entry.getValue();
+                Object value2;
+                if ((value2 = second.get(entry.getKey())) != null) {
+                    if (value2.getClass() == value.getClass()) {
+                        if (value instanceof byte[]) {
+                            if (!Arrays.equals((byte[])value, (byte[])value2)) {
+                                return false;
+                            }
+                        } else if (value instanceof Map) {
+                            //noinspection unchecked
+                            if (!testSourceMapEquality((Map<String, Object>)value, (Map<String, Object>)value2)) {
+                                return false;
+                            }
+                        } else if (!value.equals(value2)) {
+                            return false;
+                        }
+                    }
+                } else {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    @Override
+    protected WriteableIngestDocument createTestInstance() {
+        // We want to create a test instance without byte arrays. For testing byte arrays see testSourceFromXContentWithByteArray.
+        // This is needed because for comparing byte[] Object.equal fails
+        IngestDocument ingestDocument = RandomDocumentPicks.randomIngestDocument(random(), false);
+        return new WriteableIngestDocument(new IngestDocument(ingestDocument));
+    }
+
+    @Override
+    protected WriteableIngestDocument doParseInstance(XContentParser parser) {
+        return WriteableIngestDocument.fromXContent(parser);
+    }
+
+    @Override
+    protected boolean supportsUnknownFields() {
+        // Cannot support unknown fields because equality changes if new keys are added to _source
```

Comment:
good, that makes sense

Reply:
to clarify, this is because _source is mixed up with metadata right? if it was in a separate field, we would exclude it and we could return true here?

Reply:
We can do that I think. Since the metadata is separate. We would also need to ignore the `_ingest` field.